app/src/main/java/tech/relaycorp/gateway/pdc/local/PDCServer.kt

```diff
@@ -1,18 +1,54 @@
 package tech.relaycorp.gateway.pdc.local
 
-import io.ktor.application.Application
 import io.ktor.application.install
+import io.ktor.routing.Routing
 import io.ktor.routing.routing
+import io.ktor.server.engine.embeddedServer
+import io.ktor.server.netty.Netty
 import io.ktor.websocket.WebSockets
-import tech.relaycorp.gateway.domain.endpoint.EndpointRegistration
-import tech.relaycorp.gateway.pdc.local.routes.registerEndpointRegistration
-import tech.relaycorp.gateway.pdc.local.routes.registerParcelCollection
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import tech.relaycorp.gateway.pdc.local.routes.EndpointRegistrationRoute
+import tech.relaycorp.gateway.pdc.local.routes.PDCServerRoute
+import tech.relaycorp.gateway.pdc.local.routes.ParcelCollectionRoute
+import javax.inject.Inject
+import kotlin.time.seconds
 
-fun Application.configure(endpointRegistration: EndpointRegistration) {
-    install(WebSockets)
+class PDCServer
+@Inject constructor(
+    endpointRegistrationRoute: EndpointRegistrationRoute,
+    parcelCollectionRoute: ParcelCollectionRoute
+) {
 
-    routing {
-        registerEndpointRegistration(endpointRegistration)
-        registerParcelCollection()
+    private val server by lazy {
+        embeddedServer(Netty, PORT) {
+            install(WebSockets)
+
+            routing {
+                register(endpointRegistrationRoute)
+                register(parcelCollectionRoute)
+            }
```

Comment:
From what I've seen, ktor developers encourage a pattern where you have a single function to configure the app (see: [module](https://ktor.io/servers/application.html#modules)), and that function is then called by `embeddedServer` or `withTestApplication`. I think that pattern, or something closer to it, makes sense here too: If we don't do it, we'd have to duplicate all the configuration in this file and then again in the tests (as is the case in the new `testPDCServerRoute` function).

All we're doing right now is to install WebSockets and register a couple of routes, but things are very likely to get more complex: We're likely to do `install()` more features and we're likely to do stuff inside `routing` blocks (like adding `interceptors`). [Interceptors](https://ktor.io/servers/features/routing.html#interception) is something I'm considering using so that we block web browser access across all routes.

I think having an extension function roughly analogous to the original `Application.configure()` would get around this and we'd be able to use it in this property as follows:

```
private val server by lazy {
    embeddedServer(Netty, PORT) {
        configure(arrayOf(endpointRegistrationRoute, parcelCollectionRoute))
    }
}
```

And then in the test suite as follows:

```
fun <R> testPDCServerRoute(route: PDCServerRoute, test: TestApplicationEngine.() -> R) =
    withTestApplication(
        {
            configure(arrayOf(route::register))
        },
        test
    )
```

Thoughts?

Reply:
Ok, makes sense the approach, specially if we get a lot of configuration commands. The downside is that we will be doing all that configuration always, even if it's no needed for some tests, or we might want to mock those too. But let's see first if it comes to that.